src/cli/cmd/apply/mod.rs

```diff
@@ -0,0 +1,225 @@
+}
+
+#[derive(Subcommand)]
+enum System {
+    /// Resolve the store path for a Home Manager configuration and run its activation script
+    HomeManager(HomeManager),
+
+    /// Resolve the store path for a nix-darwin configuration and run its activation script
+    NixDarwin(NixDarwin),
+
+    /// Apply the resolved store path on a NixOS system
+    #[clap(name = "nixos")]
+    NixOs(NixOs),
+}
+
+#[async_trait::async_trait]
+impl CommandExecute for ApplySubcommand {
+    async fn execute(self) -> color_eyre::Result<ExitCode> {
+        let output_ref = match &self.system {
+            System::HomeManager(home_manager) => home_manager.output_ref()?,
+            System::NixOs(nixos) => nixos.output_ref()?,
+            System::NixDarwin(nix_darwin) => nix_darwin.output_ref()?,
+        };
+
+        tracing::info!("Resolving {}", output_ref);
+
+        let output_ref = parse_flake_output_ref(&output_ref)?;
+
+        let resolved_path = FlakeHubClient::resolve(self.api_addr.as_ref(), &output_ref).await?;
+
+        tracing::debug!(
+            "Successfully resolved reference {} to path {}",
+            &output_ref,
+            &resolved_path.store_path
+        );
+
+        match self.system {
+            System::HomeManager(HomeManager { .. }) => {
+                let (profile_path, _tempdir) = apply_path_to_profile(
+                    None,
+                    &resolved_path.store_path,
+                    false, // don't sudo when running `nix build`
+                )
+                .await?;
+
+                // /nix/store/{path}/activate
+                let script_path = path!(&profile_path, HOME_MANAGER_SCRIPT);
+
+                run_script(script_path, None, HOME_MANAGER_SCRIPT).await?;
+            }
+            System::NixDarwin(NixDarwin { profile, .. }) => {
+                apply_path_to_profile(
+                    Some(&profile),
+                    &resolved_path.store_path,
+                    true, // sudo if necessary when running `nix build`
+                )
+                .await?;
+
+                // {path}/sw/bin/darwin-rebuild
+                let script_path = path!(&profile, "sw", "bin", NIX_DARWIN_SCRIPT);
+
+                run_script(
+                    script_path,
+                    Some(DARWIN_REBUILD_ACTION.to_string()),
+                    NIX_DARWIN_SCRIPT,
+                )
+                .await?;
+            }
+            System::NixOs(NixOs { action, .. }) => {
+                let profile_path = Path::new(NIXOS_PROFILE);
+                apply_path_to_profile(
+                    Some(Path::new(NIXOS_PROFILE)),
+                    &resolved_path.store_path,
+                    true, // sudo if necessary when running `nix build`
+                )
+                .await?;
+
+                let script_path = path!(&profile_path, "bin", NIXOS_SCRIPT);
+
+                run_script(script_path, Some(action.to_string()), NIXOS_SCRIPT).await?;
+            }
+        }
```

Comment:
Is there a reason this couldn't be something implemented on the inner types and passed through the enum? So it would become something like `self.system.apply_path(&resolved_path.store_path)`

Reply:
Do you mean like make a trait and use it here?

Reply:
Even with a commonly implemented trait, I'm pretty sure you'd still have to call the method on each variant, like we do with `CommandExecute`

Reply:
It doesn't have to be a trait (though that would make it slightly easier) -- just some function implemented on the `System` enum that calls into specialized functions in the inner structs.